src/bin/pgcopydb/ld_transform.c

```diff
@@ -1170,20 +1185,137 @@ parseMessage(StreamContext *privateContext, char *message, JSON_Value *json)
 
 
 /*
- * streamLogicalTransactionAppendStatement appends a statement to the current
- * transaction.
- *
- * There are two ways to append a statement to an existing transaction:
- *
- *  1. it's a new statement altogether, we just append to the linked-list
+ * coalesceLogicalTransactionStatement appends a new entry to an existing tuple
+ * array created during the last INSERT statement in a logical transaction.
  *
- *  2. it's the same statement as the previous one, we only add an entry to the
- *     already existing tuple array created on the previous statement
+ * This functionality enables the generation of multi-values INSERT or COPY
+ * commands, enhancing efficiency.
  *
- * This allows to then generate multi-values insert commands, for instance.
+ * Important: Before invoking this function, ensure that validation is performed
+ * using canCoalesceLogicalTransactionStatement.
+ */
+static bool
+coalesceLogicalTransactionStatement(LogicalTransaction *txn,
+									LogicalTransactionStatement *new)
+{
+	LogicalTransactionStatement *last = txn->last;
+
+	LogicalMessageValuesArray *lastValuesArray = &(last->stmt.insert.new.array->values);
+	LogicalMessageValuesArray *newValuesArray = &(new->stmt.insert.new.array->values);
+
+	int capacity = lastValuesArray->capacity;
+	LogicalMessageValues *array = lastValuesArray->array;
+	if (capacity < (lastValuesArray->count + 1))
+	{
+		/* double the capacity to avoid excessive realloc */
+		capacity *= 2;
+		array = (LogicalMessageValues *) realloc(array, sizeof(LogicalMessageValues) *
+												 capacity);
+		if (array == NULL)
+		{
+			log_error(ALLOCATION_FAILED_ERROR);
+			return false;
+		}
+
+		lastValuesArray->capacity = capacity;
+		lastValuesArray->array = array;
+	}
+
+	/* Move the new value from the 'newValuesArray' to the 'lastValuesArray' of
+	 * the existing statement. Additionally, set the count of the 'newValuesArray'
+	 * to 0 to prevent it from being deallocated by FreeLogicalMessageTupleArray,
+	 * as it has been moved to the 'lastValuesArray'.
+	 */
```

Comment:
Nitpicking: It would be nice to better explain the “it has been moved”, in other words recap what `realloc` is doing for us in the memory allocation areas here. Also please have the first line of the comment empty.